job-runner/chain/chain.go

```diff
@@ -292,66 +371,136 @@ func (c *chain) SetJobState(jobId string, state byte) {
 		// job must not be running.
 		delete(c.running, jobId)
 	}
+
+	if state == proto.STATE_PENDING || state == proto.STATE_STOPPED {
+		// Job was stopped or job was previously done but set back to pending
+		// (i.e. on sequence retry). Decrement Chain.numJobsRun so that # of jobs
+		// run stays correct.
+		if c.numJobsRun == 0 {
+			// Don't decrement below 0 - n is unsigned
+			return
+		}
+		c.numJobsRun--
+	}
 }
 
 // Running returns a list of running jobs.
-func (c *chain) Running() map[string]proto.JobStatus {
+func (c *Chain) Running() map[string]proto.JobStatus {
 	// Return copy of c.running
-	c.runningMux.Lock()
-	defer c.runningMux.Unlock()
+	c.runningMux.RLock()
+	defer c.runningMux.RUnlock()
 	running := make(map[string]proto.JobStatus, len(c.running))
 	for jobId, jobStatus := range c.running {
 		running[jobId] = jobStatus
 	}
 	return running
 }
 
+// Length returns the total number of jobs in the chain.
+func (c *Chain) Length() int {
+	return len(c.jobChain.Jobs)
+}
+
 // //////////////////////////////////////////////////////////////////////////
-// Implement JOSN interfaces for custom (un)marshalling by chain.Repo
+// Implement JSON interfaces for custom (un)marshalling by chain.Repo
 // //////////////////////////////////////////////////////////////////////////
 
 type chainJSON struct {
-	JobChain *proto.JobChain            `json:"jobChain"`
-	Running  map[string]proto.JobStatus `json:"running"`
-	N        uint                       `json:"n"`
+	// An SJC has already fields for most of the info we need to store.
+	proto.SuspendedJobChain
+	Running    map[string]proto.JobStatus `json:"running"`
+	NumJobsRun uint                       `json:"numJobsRun"`
 }
 
-func (c *chain) MarshalJSON() ([]byte, error) {
-	c.jcMux.Lock()
-	defer c.jcMux.Unlock()
-
-	c.runningMux.Lock()
-	defer c.runningMux.Unlock()
+func (c *Chain) MarshalJSON() ([]byte, error) {
+	c.runningMux.RLock()
+	running := c.running
+	numJobsRun := c.numJobsRun
+	c.runningMux.RUnlock()
 
 	m := chainJSON{
-		JobChain: c.jobChain,
-		Running:  c.running,
-		N:        c.n,
+		SuspendedJobChain: c.ToSuspended(),
```

Comment:
I'm a little confused as to why whenever you marshal a job chain you will get a suspended job chain. Maybe a comment would help

Reply:
Also, in a follow up PR, we should probably get rid of the chain.RedisRepo (we've never actually used it). It marshals the Chain struct when you add a Chain to the repo, and I'm not sure if turning the Chain into a suspended Chain every time that happens would have unintended consequences.

Reply:
This was just a sort of convenience, since chain.ToSuspended() already saves most of the things you need to know to recreate a Chain. I see how it's confusing to "suspend" the chain whenever you marshal it, though. I'll change this so it doesn't use an SJC or ToSuspended() anymore.

As for the chain repo, we can either get rid of it somewhen or adapt it as a way to recover job chains after Job Runner crashes.